### PR TITLE
[SPARK-30613][SQL] Support Hive style REPLACE COLUMNS syntax

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

This PR proposes to support Hive-style `ALTER TABLE ... REPLACE COLUMNS ...` as described in https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-Add/ReplaceColumns

The user now can do the following:
```SQL
CREATE TABLE t (col1 int, col2 int) USING Foo;
ALTER TABLE t REPLACE COLUMNS (col2 string COMMENT 'comment2', col3 int COMMENT 'comment3');
```
, which drops the existing columns `col1` and `col2`, and add new columns `col2` and `col3`.

### Why are the changes needed?

This is a new DDL statement. Spark currently supports the Hive-style `ALTER TABLE ... CHANGE COLUMN ...`, so this new addition can be useful.

### Does this PR introduce any user-facing change?

Yes, adding a new DDL statement.

### How was this patch tested?

More tests to be added.

### CODE REVIEW

**Comment by imback82** [comment]
@cloud-fan This is WIP, but I have a couple of questions.

`REPLACE COLUMNS` needs to drop all the existing columns, so I am creating `Seq[TableChange]` which has `DeleteColumn`s followed by `AddColumn`s.

1) Can we assume that `TableCatalog.alterTable()` would apply the changes in the given order? (this is not documented).
2) Since it needs to drop all the existing columns, we need to look up the table before creating `AlterTable` logical plan. What I currently have is to call `loadTable` in `ResolveCatalogs`, which may not be ideal since we will do two look ups (another in `ResolveTables`). Another way is to register a callback to `AlterTable` which can be called after table is resolved. What do you think?

**Comment by cloud-fan** [comment]
> apply the changes in the given order?

This is a good point. I think we should, can you open a PR to improve the doc?

> which may not be ideal since we will do two look ups 

I think it's OK. We can clean it up later, thinking about how to resolve commands in general.

**Comment by imback82** [comment]
@cloud-fan this is now ready for review. Thanks!

**Comment by imback82** [comment]
We currently have the following for `ADD COLUMN`
```
    | ALTER TABLE multipartIdentifier
        ADD (COLUMN | COLUMNS)
        columns=qualifiedColTypeWithPositionList                       #addTableColumns
    | ALTER TABLE multipartIdentifier
        ADD (COLUMN | COLUMNS)
        '(' columns=qualifiedColTypeWithPositionList ')'               #addTableColumns
```

But it seems that only the following is the sql standard:
```
    | ALTER TABLE multipartIdentifier
        ADD COLUMN?
        column=qualifiedColTypeWithPosition 
```
, and the following is Hive style:
```
    | ALTER TABLE multipartIdentifier
        ADD COLUMNS
        '(' columns=qualifiedColTypeWithPositionList ')' 
```
Should we fix this as well? (if so, we can combine hive style ADD and REPLACE grammar easily as well.)

**Comment by cloud-fan** [comment]
The problem is that we can't remove a SQL syntax that works in prior releases. Maybe we have to bear with it here.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala [hunk 1]
```diff
@@ -1073,4 +1073,20 @@ trait AlterTableTests extends SharedSparkSession {
       assert(updated.properties === withDefaultOwnership(Map("provider" -> v2Format)).asJava)
     }
   }
+
+  test("AlterTable: replace columns") {
+    val t = s"${catalogAndNamespace}table_name"
+    withTable(t) {
+      sql(s"CREATE TABLE $t (col1 int, col2 int) USING $v2Format")
+      sql(s"ALTER TABLE $t REPLACE COLUMNS " +
+        "(col2 string COMMENT 'comment2', col3 int COMMENT 'comment3')")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala]: One question: if the `col2` already has comment but we don't specify new comment in `REPLACE COLUMNS`, will we retain the comment?

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala (hunk 1 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/AlterTableTests.scala]: The behavior of `REPLACE COLUMNS` is to drop all the existing columns first then add new columns. Thus, the comment will not be retained. I will update the test to reflect this.

**Comment by cloud-fan** [comment]
LGTM, merging to master!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -470,9 +470,15 @@ trait CheckAnalysis extends PredicateHelper {
               }
             }
 
+            val colsToDelete = mutable.Set.empty[Seq[String]]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: This causes conflicts when I backport https://github.com/apache/spark/pull/27584

I think the change in this file should go into 3.0 as well. Logically columns deleted should be skipped when checking name duplication for AddColumn.

@imback82 can you open a PR to backport https://github.com/apache/spark/pull/27584 with changes in this file?

Comment (imback82) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: Yes, working on it now!

**Pull request closed without merging**